workflow/executor/executor.go

```diff
@@ -931,6 +958,100 @@ func (we *WorkflowExecutor) Wait(ctx context.Context) error {
 	return nil
 }
 
+var progressRegexp = regexp.MustCompile("#argo .*progress=([0-9]+/[0-9]+)")
+
+// monitorProgress monitors for self-reported progress in the progressFile and patches the pod annotations with the parsed progress.
+//
+// The function reads the last line of the `progressFile` every `readFileTickDuration`.
+// If the line matches `#argo progress=N/M`, will set the progress annotation to the parsed progress value.
+// Every `annotationPatchTickDuration` the pod is patched with the updated annotations. This way the controller
+// gets notified of new self reported progress.
+func (we *WorkflowExecutor) monitorProgress(ctx context.Context, progressFile string) {
+	annotationPatchTicker := time.NewTicker(we.annotationPatchTickDuration)
+	defer annotationPatchTicker.Stop()
+	fileTicker := time.NewTicker(we.readProgressFileTickDuration)
+	defer fileTicker.Stop()
+
+	lastLine := ""
+	progressFile = filepath.Clean(progressFile)
+	// if progress has not been detected once within 10 ticks, abort monitoring
+	countDown := 10
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.WithError(ctx.Err()).Info("context done")
+			return
+		case <-annotationPatchTicker.C:
+			log.WithField("annotations", we.annotations).Infof("patching pod annotations")
+			err := we.patchPodAnnotations(ctx)
+			if err != nil {
+				log.WithError(err).Warn("failed to patch progress annotations")
+			}
+		case <-fileTicker.C:
+			if !we.isRunning(ctx) {
+				log.Info("pod not running yet, skipping tick")
+				continue
+			}
+
+			if countDown == 0 {
```

Comment:
why do you think we need this? hard to test, could we eschew it?

Reply:
I think if the workflow does not report it, it'll still work without this?

Reply:
yes, it would. Without this check there is slightly more overhead during the rest of the workflows' duration due to the monitor progress. As that shouldn't be too much I'm fine with removing this :+1: can always re-add if needed, I guess.